spec/assoc.test.sh

```diff
@@ -812,3 +812,24 @@ orange is orange
 lemon is yellow
 banana is yellow
 ## END
+
+
+#### BashAssoc ${a[@]@Q}
+
+declare -A a=()
+a['symbol1']=\'\'
+a['symbol2']='"'
+a['symbol3']='()<>&|'
+a['symbol4']='[]*?'
+echo "[${a[@]@Q}]"
+echo "[${a[*]@Q}]"
+
+## STDOUT:
+[$'\'\'' '"' '()<>&|' '[]*?']
+[$'\'\'' '"' '()<>&|' '[]*?']
+## END
+
+## OK bash STDOUT:
+['[]*?' ''\'''\''' '"' '()<>&|']
+['[]*?' ''\'''\''' '"' '()<>&|']
```

Comment:
hm bash's quoting is very odd!   It contains extra empty strings

It could just be `\'\'`

Reply:
I thought you knew this simple quoting method. For an arbitrary string, one can single-quote it by replacing all the occurrences of `'` to `'\''` and surrounding it by `'...'`. To express the algorithm in a shell script, for an arbitrary string `$str`, one can generate a quoted value by `quoted="'${str//$q/$Q}'"`, where `q=\' Q="'\''"`. Didn't know this?

Reply:
Oh yes that's true ... I did know that idiom, but I didn't recognize it there

So yes it is a correct and simple algorithm, but I guess this is one of the worst cases for it

---

In Oils we usually prefer to print the string on a single line, even if it contains `\n`, which means you have to use the `$''` kind of string ...   We might want more control over that, in people want to print POSIX strings, but so far there doesn't seem like demand, and bash sometimes uses `$''` strings as well

Reply:
> Oh yes that's true ... I did know that idiom, but I didn't recognize it there

Ah, yeah, It's not so obvious that this is generated by that idiom. I probably got used to it too much so that I can immediately recognize it subconsciously.

> (In Oils we usually prefer to print the string on a single line, even if it contains `\n`, which means you have to use the `$''` kind of string ...

Yeah, I was searching QSN to check whether you have mentioned the idiom in some of the pages of QSN.

Reply:
Oh "QSN" is now "J8 Notation", although they are virtually the same for strings, it's C-style `\`  escapes, which POSIX shell doesn't have

I think Oils could better support/document the pure POSIX idiom -- we probably should have a builtin function, even though it is trivial to write it in OSH/YSH 

The idiom produces POSIX shell strings, but not J8 strings, which don't have the "word joining"

(Actually "word joining" like `foo\'bar` in YSH may be disallowed, except in limited cases like `--foo='bar'` ... basically because there are so many ways of quoting, most users aren't used to reading it, and C-style quoting is powerful to express every string)